Service Catalog Provisioned Product and related resources

ahgittin opened the pull request:
This PR adds four resources:

* `servicecatalog_portfolio_principal_association`
* `resource_aws_servicecatalog_portfolio_product_association`
* `resource_aws_servicecatalog_product`
* `resource_aws_servicecatalog_provisioned_product`

Our goal was just to add `provisioned_product` -- so that Terraform can provision things from the AWS Service Catalog.

However in order to have a test which runs and doesn't require custom AWS setup, we needed `portfolio` (which existed and worked well), `product` (where most of the work was already done in https://github.com/terraform-providers/terraform-provider-aws/pull/4980, included here with minor fixes) , and the two `porfolio-{principal,product}-associations` (because in order to provision a product, it must be associated to a portfolio, and the portfolio must be associated to a principal). 
 Thus please forgive the size of this PR.  And we are new contributors to this codebase so please also forgive any deviations from best practices!

This has tests for each -- the test for product creates all the pre-req objects noted above and then successfully provisions the product.  We also have the markdown documentation for each.  If there is anything we've omitted let us know.

We have time blocked out this week and next to correct/improve anything y'all spot.

We also intend to add the remaining Service Catalog resources (8 left) as we use quite a few of them. 
 We'll do this in a follow-on PR.  So any tips are welcome, even things that aren't necessary to address in this PR.

Thanks.

DrFaust92 commented on aws/resource_aws_servicecatalog_portfolio_principal_association.go:
```diff
@@ -0,0 +1,165 @@
+	input := servicecatalog.AssociatePrincipalWithPortfolioInput{
+		PortfolioId:   aws.String(portfolioId),
+		PrincipalARN:  aws.String(principalArn),
+		PrincipalType: aws.String("IAM"),
```
can you change hardcoded "IAM" to `servicecatalog.PrincipalTypeIam`

ahgittin replied:
done (check-in to come)

DrFaust92 commented on aws/resource_aws_servicecatalog_portfolio_principal_association.go:
```diff
@@ -0,0 +1,165 @@
+			},
+			"principal_arn": {
+				Type:     schema.TypeString,
+				Required: true,
```
please add validation here using `ValidateFunc: validateArn,`

ahgittin replied:
done (check-in to come)

DrFaust92 commented on aws/resource_aws_servicecatalog_portfolio_principal_association.go:
```diff
@@ -0,0 +1,165 @@
+package aws
+
+import (
+	"fmt"
```
small nit, can you split the deps to go packages and external dep packages like other resources?

ahgittin replied:
done - all files (check-in to come)

DrFaust92 commented on aws/resource_aws_servicecatalog_portfolio_principal_association.go:
```diff
@@ -0,0 +1,165 @@
+			return err
+		}
+		for _, principal := range pageOfDetails {
+			if *principal.PrincipalARN == principalArn {
```
wrap `principal.PrincipalARN` with `aws.StringValue()`

ahgittin replied:
done (check-in to come)

DrFaust92 commented on aws/resource_aws_servicecatalog_portfolio_principal_association.go:
```diff
@@ -0,0 +1,165 @@
+		if nextPageToken == nil || isFound {
+			break
+		}
+		pageToken = *nextPageToken
```
same as  above

ahgittin replied:
done (check-in to come)

DrFaust92 commented on aws/resource_aws_servicecatalog_portfolio_principal_association.go:
```diff
@@ -0,0 +1,165 @@
+func resourceAwsServiceCatalogPortfolioPrincipalAssociationUpdate(d *schema.ResourceData, meta interface{}) error {
+	const principalArnKey = "principal_arn"
+	const portfolioIdKey = "portfolio_id"
+	if d.HasChange(principalArnKey) || d.HasChange(portfolioIdKey) {
```
change to a single `d.HasChanges(principalArnKey, portfolioIdKey)`

DrFaust92 replied:
btw, any reason for the consts? using the string literals are the way its done in other resources

ahgittin replied:
the linter picked these up and already changed (already checked in)

the only reason for the constants was that in some projects that's preferred (to prevent typos) but we've switched to this project's convention of repreated string literals

DrFaust92 commented on aws/resource_aws_servicecatalog_portfolio_principal_association.go:
```diff
@@ -0,0 +1,163 @@
+		oldPortfolioId, newPortfolioId := d.GetChange("portfolio_id")
+		d.Set("principal_arn", oldPrincipalArn.(string))
+		d.Set("portfolio_id", oldPortfolioId.(string))
+		resourceAwsServiceCatalogPortfolioPrincipalAssociationDelete(d, meta)
```
this is a bit unusual as far as i can tell, same thing for the call to create. can we extract the logic to other functions and re use them? update should only call read

ahgittin replied:
Unsure about this one:

The only way to "update" this resource is to delete the old and create a new.  So this implementation seemed a good way.  But happy to change it.  Options I see are:

(a) As is -- call the canonical "delete" and "create" methods
(b) Refactor the common delete (and create) logic eg to a `doDelete` (and `doCreate`) which gets used by the canonical "delete" ("create") and "update"
(c) Just duplicate the delete and create logic here
(d) Mark the fields as `ForceNew: true` and make this method return an `UnexpectedInvocation` error

What do you recommend?

ahgittin replied:
(d) seems the right way:  as soon as all fields are `ForceNew` it complains if the `update` method even exists (nice)

DrFaust92 commented on aws/resource_aws_servicecatalog_portfolio_product_association.go:
```diff
@@ -0,0 +1,143 @@
+			return err
+		}
+		for _, portfolioDetail := range pageOfDetails {
+			if *portfolioDetail.Id == portfolioId {
```
warp `portfolioDetail.Id` with `aws.StringValue()`

ahgittin replied:
done (check-in to come)

DrFaust92 commented on aws/resource_aws_servicecatalog_portfolio_product_association.go:
```diff
@@ -0,0 +1,143 @@
+		if nextPageToken == nil || isFound {
+			break
+		}
+		pageToken = *nextPageToken
```
same as above

ahgittin replied:
done (check-in to come)

DrFaust92 commented on aws/resource_aws_servicecatalog_product_test.go:
```diff
@@ -0,0 +1,241 @@
+			{
+				Config: testAccCheckAwsServiceCatalogProductResourceConfigTemplate(arbitraryBucketName, arbitraryProductName, arbitraryProvisionArtifactName, tag1, tag2),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("aws_servicecatalog_product.test", "name", arbitraryProductName),
```
small nit, can you extract this to a variable and define it once?
`resourceName := "aws_servicecatalog_product.test"` and pass `resourceName` to to other funcs

ahgittin replied:
done (check-in to come)

DrFaust92 commented on aws/resource_aws_servicecatalog_product_test.go:
```diff
@@ -0,0 +1,241 @@
+					resource.TestCheckResourceAttrSet("aws_servicecatalog_product.test", "support_description"),
+					resource.TestCheckResourceAttrSet("aws_servicecatalog_product.test", "support_email"),
+					resource.TestCheckResourceAttrSet("aws_servicecatalog_product.test", "support_url"),
+				),
```
add import check

DrFaust92 replied:
this relevant for all tests as well

DrFaust92 commented on aws/resource_aws_servicecatalog_product_test.go:
```diff
@@ -0,0 +1,241 @@
+					resource.TestCheckResourceAttr("aws_servicecatalog_product.test", "provisioning_artifact.#", "1"),
+					resource.TestCheckResourceAttr("aws_servicecatalog_product.test", "provisioning_artifact.0.name", arbitraryProvisionArtifactName),
+					resource.TestCheckResourceAttrSet("aws_servicecatalog_product.test", "provisioning_artifact.0.description"),
+					resource.TestCheckResourceAttr("aws_servicecatalog_product.test", "tags.%", "2"),
```
remove tags from basic config and check for `tags.%` is `0`

DrFaust92 commented on aws/resource_aws_servicecatalog_product_test.go:
```diff
@@ -0,0 +1,241 @@
+	})
+}
+
+func TestAccAWSServiceCatalogProduct_updateTags(t *testing.T) {
```
can you rewrite this based on https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/contribution-checklists.md#resource-tagging-acceptance-testing-implementation

DrFaust92 commented on aws/resource_aws_servicecatalog_provisioned_product.go:
```diff
@@ -0,0 +1,370 @@
+				Type:     schema.TypeList,
+				ForceNew: true,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
```
add `ValidateFunc: validateArn,`

ahgittin replied:
done (check-in to come)

DrFaust92 commented on aws/resource_aws_servicecatalog_provisioned_product.go:
```diff
@@ -0,0 +1,370 @@
+	resp, err := conn.DescribeProvisionedProduct(&input)
+
+	if err != nil {
+		if scErr, ok := err.(awserr.Error); ok && scErr.Code() == "ResourceNotFoundException" {
```
replace with `isAWSErr(err, servicecatalog.ErrCodeResourceNotFoundException, "")

ahgittin replied:
done (check-in to come)

ahgittin commented:
Thanks @DrFaust92 -- all comments make sense.  I've addressed most and pushed those.

Remaining items are:

* Calling `delete` and `create` during `update` for the two associations.  We have *one open question* about the best way to resolve -- 
https://github.com/terraform-providers/terraform-provider-aws/pull/13797#discussion_r442055262

* Expand acceptance tests -- thank you for the useful link

* Split into multiple PRs -- will do.  Please see the proposal below.

For splitting this into multiple PRs -- we wanted to avoid merge conflicts in the indexes (`website/aws.erb` and `provider.go` files) and there are some test dependencies between the resources, so what we propose is:
  * new PR for just `product` -- but not updating the indexes yet
  * new PR for the `portfolio-principal association` -- no dependencies, not updating the indexes yet
  * new PR for the `portfolio-product association` -- test would depend on `product`, so would include that; but not updating the indexes yet
  * this PR will be for `provisioned_product` -- test depends on all the above, so include those PRs, and now updating the indexes
We also want to keep the commit history (so the people who did the work are acknowledged) so we propose adding new commits to remove the irrelevant files, which we'll include but revert in this PR.  This way each PR will be small deltas on prior ones so easier to review -- but we won't get conflicts and we keep in the history the commits that different contributors have done.  (Happy to do it however is best for you all.)

ahgittin commented:
^ the attempt to leave out the change to index file `provider.go` of course doesn't work because the tests don't run. we will open multiple PRs one per resource and manually resolve the likely (but trivial) merge conflicts in `provider.go` and `website/aws.erb`

ahgittin commented:
I've opened #13834 containing just the `aws_servicecatalog_product` code, with tests updated as per your comments here.

I will do the same for the two associations, and then use this PR for the `provisioned_product` which of course depends on the others.  Until that is ready I will mark this [WIP].

ahgittin commented:
update - this PR is still WIP, working on a test for the parameters to a provisioned product.

however the other PRs are all created and ready to go:

* #13834 - SC Product
* #13837 - SC PortfolioPrincipalAssociation
* #13864 - SC PortfolioProductAssociation (builds on SC Product)
* #13866 - SC Portfolio (minor update to existing)

The delta between those and this will be just what is needed for the ProvisionedProduct (which is the resource we are mainly interested in).

cc @DrFaust92 - are we doing it right?  i saw a wee 👍 at one point so I hope so!

DrFaust92 commented:
@ahgittin, split looks good I'll take a closer look later today or in the next couple of days and give see feedback to help ease the work for maintaners.

ahgittin commented:
Test with parameters, outputs, and non-force-new update now all working.

I think this is ready for review (after the other PRs noted above).

We have one difficulty with the import test, as noted in the provisioned_product_test -- it doesn't seem to be compatible with the custom provider assuming a custom role in this case.  So the import test section is commented out, with the reason why.  I welcome any advice on ways around this (separate test configurations?) or maybe it could be accepted without it.

Aside from that there are a few TODOs for things that could be added in future (support for budgets, stack set preferences) but aren't necessary for most functionality.

Acceptance test output for the `ProvisionedProduct` tests:

```
make testacc TESTARGS='-run=TestAccAWSServiceCatalogProvisionedProduct' 2>&1 | tee x.log                           23:35:33 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSServiceCatalogProvisionedProduct -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSServiceCatalogProvisionedProduct_basic
=== PAUSE TestAccAWSServiceCatalogProvisionedProduct_basic
=== RUN   TestAccAWSServiceCatalogProvisionedProduct_disappears
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_provisioned_product.test (prepare state)
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node aws_iam_role_policy.test (prepare state)
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/22 23:40:42 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test (prepare state)
--- PASS: TestAccAWSServiceCatalogProvisionedProduct_disappears (151.98s)
=== RUN   TestAccAWSServiceCatalogProvisionedProduct_tags
2020/06/22 23:42:35 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/22 23:42:35 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/22 23:42:35 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/22 23:42:35 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/22 23:44:13 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/22 23:44:13 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/22 23:44:13 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/22 23:44:13 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_provisioned_product.test (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node aws_iam_role_policy.test (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/22 23:45:49 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
--- PASS: TestAccAWSServiceCatalogProvisionedProduct_tags (344.46s)
=== RUN   TestAccAWSServiceCatalogProvisionedProduct_ProvisioningParameters
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_iam_role_policy.test (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template_params (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test_params (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test_params (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_provisioned_product.test_params (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/22 23:49:22 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
--- PASS: TestAccAWSServiceCatalogProvisionedProduct_ProvisioningParameters (210.52s)
=== CONT  TestAccAWSServiceCatalogProvisionedProduct_basic
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node aws_iam_role_policy.test (prepare state)
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test (prepare state)
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_provisioned_product.test (prepare state)
2020/06/22 23:51:45 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
--- PASS: TestAccAWSServiceCatalogProvisionedProduct_basic (144.94s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	853.865s
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/flatmap	0.820s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags	0.338s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/naming	0.333s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/apigatewayv2/waiter	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/batch/equivalency	0.247s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ecs/waiter	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/eks/token	0.393s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/guardduty/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/kinesisanalytics/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/kms/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/neptune/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/rds/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/secretsmanager/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/servicediscovery/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/workspaces/waiter	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/tfawsresource	0.532s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/awsproviderlint	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/helper/awsprovidertype/keyvaluetags	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes	0.767s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSAT001	1.127s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSR001	0.478s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSR002	0.976s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/fmtsprintfcallexpr	0.214s [no tests to run]
```

ahgittin commented:
And for completeness the acceptance tests of all Service Catalog resources thus far -- including those on the PR's listed above and here:

```
$ make testacc TESTARGS='-run=TestAccAWSServiceCatalog'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSServiceCatalog -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSServiceCatalogPortfolioPrincipalAssociation_basic
=== PAUSE TestAccAWSServiceCatalogPortfolioPrincipalAssociation_basic
=== RUN   TestAccAWSServiceCatalogPortfolioPrincipalAssociation_disappears
=== PAUSE TestAccAWSServiceCatalogPortfolioPrincipalAssociation_disappears
=== RUN   TestAccAWSServiceCatalogPortfolioPrincipalAssociation_Portfolio_update
=== PAUSE TestAccAWSServiceCatalogPortfolioPrincipalAssociation_Portfolio_update
=== RUN   TestAccAWSServiceCatalogPortfolioPrincipalAssociation_Principal_update
=== PAUSE TestAccAWSServiceCatalogPortfolioPrincipalAssociation_Principal_update
=== RUN   TestAccAWSServiceCatalogPortfolioPrincipalAssociation_update_all
=== PAUSE TestAccAWSServiceCatalogPortfolioPrincipalAssociation_update_all
=== RUN   TestAccAWSServiceCatalogPortfolioProductAssociation_basic
=== PAUSE TestAccAWSServiceCatalogPortfolioProductAssociation_basic
=== RUN   TestAccAWSServiceCatalogPortfolioProductAssociation_disappears
=== PAUSE TestAccAWSServiceCatalogPortfolioProductAssociation_disappears
=== RUN   TestAccAWSServiceCatalogPortfolioProductAssociation_Portfolio_update
=== PAUSE TestAccAWSServiceCatalogPortfolioProductAssociation_Portfolio_update
=== RUN   TestAccAWSServiceCatalogPortfolioProductAssociation_Product_update
=== PAUSE TestAccAWSServiceCatalogPortfolioProductAssociation_Product_update
=== RUN   TestAccAWSServiceCatalogPortfolioProductAssociation_update_all
=== PAUSE TestAccAWSServiceCatalogPortfolioProductAssociation_update_all
=== RUN   TestAccAWSServiceCatalogPortfolio_basic
=== PAUSE TestAccAWSServiceCatalogPortfolio_basic
=== RUN   TestAccAWSServiceCatalogPortfolio_disappears
=== PAUSE TestAccAWSServiceCatalogPortfolio_disappears
=== RUN   TestAccAWSServiceCatalogPortfolio_tags
=== PAUSE TestAccAWSServiceCatalogPortfolio_tags
=== RUN   TestAccAWSServiceCatalogProduct_basic
--- PASS: TestAccAWSServiceCatalogProduct_basic (34.69s)
=== RUN   TestAccAWSServiceCatalogProduct_disappears
2020/06/23 00:09:36 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/23 00:09:36 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:09:36 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:09:36 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
--- PASS: TestAccAWSServiceCatalogProduct_disappears (31.98s)
=== RUN   TestAccAWSServiceCatalogProduct_tags
2020/06/23 00:10:46 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:10:46 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/23 00:10:46 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:10:46 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
--- PASS: TestAccAWSServiceCatalogProduct_tags (71.48s)
=== RUN   TestAccAWSServiceCatalogProduct_ProvisioningArtifact_updateInPlace
2020/06/23 00:11:39 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:11:39 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/23 00:11:39 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/23 00:11:39 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
--- PASS: TestAccAWSServiceCatalogProduct_ProvisioningArtifact_updateInPlace (52.60s)
=== RUN   TestAccAWSServiceCatalogProduct_Bucket_updateForcesNew
2020/06/23 00:12:37 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:12:37 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:12:37 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/23 00:12:37 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
--- PASS: TestAccAWSServiceCatalogProduct_Bucket_updateForcesNew (58.24s)
=== RUN   TestAccAWSServiceCatalogProvisionedProduct_basic
=== PAUSE TestAccAWSServiceCatalogProvisionedProduct_basic
=== RUN   TestAccAWSServiceCatalogProvisionedProduct_disappears
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test (prepare state)
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_provisioned_product.test (prepare state)
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node aws_iam_role_policy.test (prepare state)
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/23 00:15:08 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
--- PASS: TestAccAWSServiceCatalogProvisionedProduct_disappears (150.69s)
=== RUN   TestAccAWSServiceCatalogProvisionedProduct_tags
2020/06/23 00:16:59 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/23 00:16:59 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/23 00:16:59 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:16:59 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:18:38 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:18:38 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/23 00:18:38 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:18:38 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_provisioned_product.test (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node aws_iam_role_policy.test (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/23 00:20:13 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
--- PASS: TestAccAWSServiceCatalogProvisionedProduct_tags (343.44s)
=== RUN   TestAccAWSServiceCatalogProvisionedProduct_ProvisioningParameters
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_iam_role_policy.test (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_provisioned_product.test_params (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test_params (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template_params (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test_params (prepare state)
2020/06/23 00:23:56 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
--- PASS: TestAccAWSServiceCatalogProvisionedProduct_ProvisioningParameters (222.66s)
=== CONT  TestAccAWSServiceCatalogPortfolioPrincipalAssociation_basic
=== CONT  TestAccAWSServiceCatalogPortfolioProductAssociation_Product_update
=== CONT  TestAccAWSServiceCatalogProvisionedProduct_basic
=== CONT  TestAccAWSServiceCatalogPortfolioProductAssociation_disappears
=== CONT  TestAccAWSServiceCatalogPortfolioPrincipalAssociation_Portfolio_update
=== CONT  TestAccAWSServiceCatalogPortfolioProductAssociation_basic
=== CONT  TestAccAWSServiceCatalogPortfolioPrincipalAssociation_disappears
=== CONT  TestAccAWSServiceCatalogPortfolio_tags
=== CONT  TestAccAWSServiceCatalogPortfolio_disappears
=== CONT  TestAccAWSServiceCatalogPortfolioPrincipalAssociation_Principal_update
=== CONT  TestAccAWSServiceCatalogPortfolio_basic
=== CONT  TestAccAWSServiceCatalogPortfolioProductAssociation_update_all
=== CONT  TestAccAWSServiceCatalogPortfolioPrincipalAssociation_update_all
=== CONT  TestAccAWSServiceCatalogPortfolioProductAssociation_Portfolio_update
--- PASS: TestAccAWSServiceCatalogPortfolio_disappears (15.31s)
--- PASS: TestAccAWSServiceCatalogPortfolio_basic (20.05s)
2020/06/23 00:25:12 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
2020/06/23 00:25:12 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test-20i2k (prepare state)
2020/06/23 00:25:12 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test-20i2k (prepare state)
2020/06/23 00:25:13 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:25:13 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/23 00:25:13 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test-02ckv (prepare state)
2020/06/23 00:25:13 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:25:13 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test-02ckv (prepare state)
2020/06/23 00:25:13 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
--- PASS: TestAccAWSServiceCatalogPortfolioPrincipalAssociation_disappears (35.14s)
2020/06/23 00:25:14 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:25:14 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:25:14 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
--- PASS: TestAccAWSServiceCatalogPortfolioPrincipalAssociation_basic (37.73s)
--- PASS: TestAccAWSServiceCatalogPortfolioProductAssociation_disappears (38.35s)
--- PASS: TestAccAWSServiceCatalogPortfolioProductAssociation_basic (38.74s)
2020/06/23 00:25:22 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
--- PASS: TestAccAWSServiceCatalogPortfolio_tags (44.84s)
2020/06/23 00:25:37 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test-9grkg (prepare state)
2020/06/23 00:25:37 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:25:37 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:25:37 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test-q32gf (prepare state)
2020/06/23 00:25:37 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/23 00:25:37 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/23 00:25:37 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
2020/06/23 00:25:37 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test-hdmr0 (prepare state)
2020/06/23 00:25:37 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test-zke10 (prepare state)
2020/06/23 00:25:39 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test-q6jvl (prepare state)
2020/06/23 00:25:39 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
2020/06/23 00:25:39 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test-q6jvl (prepare state)
--- PASS: TestAccAWSServiceCatalogPortfolioPrincipalAssociation_Portfolio_update (62.06s)
--- PASS: TestAccAWSServiceCatalogPortfolioProductAssociation_Portfolio_update (62.40s)
--- PASS: TestAccAWSServiceCatalogPortfolioPrincipalAssociation_update_all (63.10s)
2020/06/23 00:25:42 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test-ij3ke (prepare state)
2020/06/23 00:25:42 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:25:42 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:25:42 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test-ij3ke (prepare state)
2020/06/23 00:25:42 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/23 00:25:42 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/23 00:25:44 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:25:44 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test-ywrgl (prepare state)
2020/06/23 00:25:44 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test-wydjj (prepare state)
2020/06/23 00:25:44 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
2020/06/23 00:25:44 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:25:44 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/23 00:25:45 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test-4he72 (prepare state)
2020/06/23 00:25:45 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test-dhp12 (prepare state)
2020/06/23 00:25:45 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
--- PASS: TestAccAWSServiceCatalogPortfolioProductAssociation_update_all (68.16s)
--- PASS: TestAccAWSServiceCatalogPortfolioPrincipalAssociation_Principal_update (69.08s)
--- PASS: TestAccAWSServiceCatalogPortfolioProductAssociation_Product_update (70.22s)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio.test (prepare state)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_product.test (prepare state)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_provisioned_product.test (prepare state)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket_object.template1 (prepare state)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node aws_s3_bucket.bucket (prepare state)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node aws_iam_role.test (prepare state)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node aws_iam_role_policy.test (prepare state)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_principal_association.test (prepare state)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node aws_servicecatalog_portfolio_product_association.test (prepare state)
2020/06/23 00:26:28 DestroyEdgeTransformer: pruning unused resource node data.aws_region.current (prepare state)
--- PASS: TestAccAWSServiceCatalogProvisionedProduct_basic (151.31s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1118.458s
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/flatmap	1.173s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags	0.286s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/naming	0.744s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/apigatewayv2/waiter	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/batch/equivalency	0.364s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ecs/waiter	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/eks/token	0.518s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/guardduty/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/kinesisanalytics/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/kms/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/neptune/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/rds/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/secretsmanager/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/servicediscovery/waiter	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/aws/internal/service/workspaces/waiter	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/tfawsresource	1.772s [no tests to run]
?   	github.com/terraform-providers/terraform-provider-aws/awsproviderlint	[no test files]
?   	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/helper/awsprovidertype/keyvaluetags	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes	1.498s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSAT001	0.541s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSR001	0.198s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/AWSR002	0.327s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/awsproviderlint/passes/fmtsprintfcallexpr	0.321s [no tests to run]
```

ahgittin commented:
> @ahgittin, split looks good I'll take a closer look later today or in the next couple of days and give see feedback to help ease the work for maintaners.

@DrFaust92 and other reviewers -- anything i can do to help?  happy to make time to pair if that would speed things up on your end.

ahgittin commented:
@DrFaust92 as-polite-as-possible bump

teamterraform commented:
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

ahgittin commented:
This has been updated as per the 18 Aug notification to use the v2 sdk and remove `aws.erb` (fixing the conflicts), and also updating to conform to the stricter markdown checks.

ahgittin commented:
The `go test` failure in the checks just now appears to be an unrelated non-deterministic failure in `aws_vpc` (not touched by this work).

This PR is subsumed by https://github.com/terraform-providers/terraform-provider-aws/pull/14281 where `go test` passes even though that PR only adds new resources and tests (doesn't change these).

ahgittin commented:
Retest this please

ahgittin commented:
Closed and reopened to trigger re-testing as previous failure appeared to be unrelated.

YakDriver commented:
@ahgittin Thank you for your work on this PR! I will be working on this before long. In order to expedite the process, I will likely make some changes. Make sure that you have checked the box "Allow edits from maintainers." (Don't worry though, you will receive all credit for your contribution and code!) Also, please coordinate with us before making any commits to this branch. Again, thank you for your help and we look forward to this popular addition to the AWS provider!

YakDriver commented:
@ahgittin @cloudsoft Thank you very much for all your work on this PR! I'm excited to work with you on this. However, in order to work with this PR, I need two things:

1. You must check the "Allow edits from maintainers" box. It is currently not checked. (Only @cloudsoft will be able to check the box since the fork with this branch is in the @cloudsoft repo.)
2. Currently there are 168 commits, many of which significantly rewrite the branch. All 168 commits need to be squashed into a single commit to proceed.

Since we look forward to adding this feature to the AWS Provider, we need you to make these changes by March 11, 2021. If you are unable to do this, we understand and thank you for your work. However, we will not be able to use this contribution.

ahgittin commented:
Replaced by https://github.com/hashicorp/terraform-provider-aws/pull/18074
